### ISSUE TITLE
How to install local extensions?

### ISSUE THREAD

**Issue by arivasvera** [bitnami/bitnami-docker-magento#82]
I need to install local extensions, how can i do it this?

**Comment by andresbono** [comment]
Hi @arivasvera, what is the issue you are facing?

Is this guide useful for you? [How To Install An Extension Using Composer](https://docs.bitnami.com/aws/apps/magento/#install_extension_composer).

-Andrés B.

**Comment by arivasvera** [comment]
Hi @andresbono,
I'm testing a non-published extension.
So, I've used composer in this way:
```
composer require {vendor}/{extension}
php bin/magento module:enable {vendor_extension} --clear-static-content
php bin/magento setup:upgrade
```

but after that, after that, the site is broken. and just throw me error 500.
I tried establishing permissions to 775 for directories and 664 to files and owner to bitnami:daemon  but it still does not work

Also, I've try creating code folder in /bitnami/magento/htdocs/app, puting my extension here in {vendor}/{extension} fixing permissions and owner, trying to go to to web setup wizard->module manager. but I still getting error 500

**Comment by andresbono** [comment]
Hi @arivasvera, what do the Apache logs say?

```
/opt/bitnami/apache/logs/access_log 
/opt/bitnami/apache/logs/error_log
```

Can you find more information there?

-Andrés B.

**Comment by brunobasto** [comment]
Did you manage to fix it? I'm also wondering how to install extensions with this image.

**Comment by arivasvera** [comment]
no, I did not manage to fix it, in the end I had to use magento with a lamp installation for my purpose

**Comment by andresbono** [comment]
Hi @brunobasto, @arivasvera, could you tell us the extensions that you are trying to install without success?

**Comment by theGC** [comment]
I've been struggling to wrap my head around how we use these images with custom modules and in truth am still not exactly sure. I'm working with kubernetes via GKE & minikube locally and had potentially a similar issue, for my situation I copy the module to the code directory within our docker file:

**Dockerfile**
```
FROM bitnami/magento:2.2.5-debian-9

# add magento_code directory to the bitnami magento install
ADD ./magento_code /opt/bitnami/magento/htdocs/app/code
```

This results in an unready *pod* within kubernetes as it gives a *500* error (perhaps similar to yours)

I then log into the *pod*:

```
# kubernetes
kubectl exec -it <POD-NAME> -- /bin/bash

# or in docker
docker exec -it <CONTAINER-NAME> bash
```

and run the following:

```
# upgrade magento install to include custom module
php /opt/bitnami/magento/htdocs/bin/magento setup:upgrade

# fix permissions
chown -R bitnami:daemon /opt/bitnami/magento/htdocs
```

This for me gets things up an running but is also wrong in kubernetes. If this pod crashes and is rebuilt I lose all these changes and revert back to a broken state.

I'm going to look into altering the default **bitnami/magento > composer.json** within my **Dockerfile** as I need someway of thisoprocess being automated and actually pulling up a functional container.

**Comment by andresbono** [comment]
Hi @theGC, can you share the extension/module that you are trying to add? It would be helpful for us in order to reproduce your issue.

-Andrés B

**Comment by theGC** [comment]
@andresbono yeah no problem.

https://github.com/theGC/magento2-thebigsurf-wireupdatehook

Since posting this morning I tried adding it to packagist as well hoping it would be added on install but this doesn't appear to of worked. To do so I grabbed the default composer.json at: **/opt/bitnami/magento/htdocs/composer.json** and added my module as a rquirement:

```
{
    "name": "magento/project-community-edition",
    "description": "eCommerce Platform for Growth (Community Edition)",
    "type": "project",
    "version": "2.2.5",
    "license": [
        "OSL-3.0",
        "AFL-3.0"
    ],
    "require": {
        "magento/product-community-edition": "2.2.5",
        "composer/composer": "@alpha",
        "thebigsurf/wire-update-hook": "~1"
    },
    ...
```

Whilst the install went through fine it hadn't added the module so I'm playing around with manually adding it again which isn't ideal.

Ideally, I'd like to be able to use **Dockerfile** to get this all working, that way it should make future updates to **bitnami/magento:latest** far easier to incorporate.

**Comment by theGC** [comment]
I've had a further play with the composer version of above but it has only got me to the same stage as including the custom module within the **/opt/bitnami/magento/htdocs/app/code/** directory directly, i.e I hit a 500 error:

> Exception #0 (UnexpectedValueException): Setup version for module 'TheBigSurf_WireUpdateHook' is not specified

This obviously doesn't help too much when it comes to automating the installation of custom modules, but for the sake of completeness I was able to do this by firstly creating an **auth.json**:

```
{
    "http-basic": {
        "repo.magento.com": {
            "username": <MAGENTO_MARKETPLACE_PUBLIC_KEY>,
            "password": <MAGENTO_MARKETPLACE_PRIVATE_KEY>
        }
    }
}
``` 

You can get the public & private key values by creating a new access key within marketplace.

Once I had that I updated my **Dockerfile** to use the auth.json and require the custom module:

```
# run bitnami's magento container
FROM bitnami/magento:2.2.5-debian-9

# Require custom modules
WORKDIR /opt/bitnami/magento/htdocs/
ADD ./auth.json .
RUN composer require thebigsurf/wire-update-hook:1.0.6
```

However, as I mentioned this just gives you a 500 error. To get it working I'd need to open a shell to the container, update magento and then change the dir/file owner as above.

**Comment by andresbono** [comment]
Hi @theGC, thanks so much for sharing the solution that you have found!

**Comment by adrianbsb** [comment]
The problem lies in the file `app/etc/NonComposerComponentRegistration.php`.

In the image the folder `/app/etc` is symlinked to `/bitnami/magento/htdocs/app/etc/`, thus the before mentioned facility tries to load registration.php files from `/bitnami/magento/htdocs/app/code/*/*/registration.php`

You could create your `code` folder at `/bitnami/magento/htdocs/app/code` and your modules should be visible. However you may run into issues as [magento doesn't allow you to load templates outside of magento's root](https://github.com/magento/magento2/issues/8368#issuecomment-355147448).

**Comment by adrianbsb** [comment]
one workaround would be to use a custom NonComposerComponentRegistration.php to load modules under app/code
e.g.

```php
<?php
/**
 * Custom component registration for docker environment
 * @version 1.0
 */

//Register components (via a list of glob patterns)
namespace Magento\NonComposerComponentRegistration;

use RuntimeException;

/**
 * Include files from a list of glob patterns
 *
 * @param string $baseDir
 *
 * @throws RuntimeException
 * @return void
 */
function main($baseDir = NULL)
{
    $globPatterns = require __DIR__ . '/registration_globlist.php';
    $baseDir      = ( NULL === $baseDir ) ? dirname(dirname(__DIR__)) . '/' : $baseDir;

    foreach ($globPatterns as $globPattern) {
        // Sorting is disabled intentionally for performance improvement
        $files = glob($baseDir . $globPattern, GLOB_NOSORT);
        if ($files === false) {
            throw new RuntimeException("glob(): error with '$baseDir$globPattern'");
        }

	    array_map(function ($file) { require_once $file; }, $files);

    }
}

main('/opt/bitnami/magento/htdocs/');
```

**Comment by kevinwatson** [comment]
@adrianbsb huge thanks for providing this workaround. Where does the `NonComposerComponentRegistration.php` file go?

**Comment by Braintelligence** [comment]
So does local installation of extensions still require a workaround? How do people cope with the fact that there's almost no extensions for Magento 2.3.X available on the marketplace?

**Comment by till** [comment]
@Braintelligence define a custom repository in `composer.json` and then require/update your module through that. Or develop the module in `/app` and package it up for the marketplace/composer later.

**Comment by Braintelligence** [comment]
Sorry I left this mess already. I recommend making your own Dockerfile.

**Comment by ChSch3000** [comment]
I'm using now the Docker image from alexcheng.
https://github.com/alexcheng1982/docker-magento2
With this one I could add the module under app/code without problems.

**Comment by kgoedert** [comment]
@ChSch3000 did you find a way around it? I am having the same problem

**Comment by andresbono** [comment]
Hi @kgoedert, the last accepted solution was https://github.com/bitnami/bitnami-docker-magento/issues/82#issuecomment-412140788, but it seems to have some caveats... and it is from some time ago.

Can you tell us more details about your specific case?

**Comment by britztopher** [comment]
Same problem here when trying to install a simple [helloworld](https://www.mageplaza.com/magento-2-module-development/view-block-layout-template-magento-2.html) module to `/bitnami/magento/htdocs/app/code/`.  Whats strange is that i can enable the module and all is well, until i go to url which i get this error: 

`1 exception(s):
Exception #0 (Magento\Framework\Exception\ValidatorException): Invalid template file: '/bitnami/magento/htdocs/app/code/Mageplaza/HelloWorld/view/frontend/templates/sayhello.phtml' in module: 'Mageplaza_HelloWorld' block's name: 'helloworld_display'`

Any idea on why the magento image would state this? In addition, when i install it on my LAMP stack in aws the same module works fine

**Comment by andresbono** [comment]
Thanks for the details @britztopher. We will be investigating this on our side. In the meantime, if you have any new findings, please, share them with us.

**Comment by zahedm** [comment]
Hi,
Any update resolving this issue?

**Comment by andresbono** [comment]
Hi @zahedm,

We created an internal task to analyze it, but unfortunately, due to the fact that we are a small team and taking into account internal milestones, we couldn't give priority yet.

We will update this issue as soon as we have more news.

**Comment by marcosbc** [comment]
@zahedm I actually think this should have been solved since [`2.3.5-debian-10-r57`](https://github.com/bitnami/bitnami-docker-magento#235-debian-10-r57) as there we moved to persisting the entire Magento "htdocs" directory. I would recommend you try with 2.4.x as they are newer versions, though.

Closing this issue, but feel free to ask to re-open if you still find this happening.

NOTE: We are working on releasing a new Docker image with a re-implementation of the initialization logic. The source code will also be published now as part of the "rootfs" directory in this repo. We expect to have it released by tomorrow, you can keep an eye to [Notable Changes](https://github.com/bitnami/bitnami-docker-magento#notable-changes).

**Comment by marcosbc** [comment]
Hi all, just letting you know that we just released Magento [`2.4.1-debian-10-r80`](https://github.com/bitnami/bitnami-docker-magento#241-debian-10-r80) which now publishes the source code like I mentioned earlier. If you find any issues, let us know and we will re-open.
